whitepapers/0011_accountant.md

```diff
@@ -0,0 +1,190 @@
+
+## Overview
+
+The `global-accountant` contract on wormchain acts as an Integrity Checker. This means that Guardians will submit pre-observations to it and only finalize their observations if the `global-accountant` gives the go-ahead.
+
+`global-accountant`keeps track of the tokens locked and minted on each connected L1.
+
+Once a Token Bridge message has reached a quorum of pre-observations (i.e. 13 guardians have submitted a pre-observation), it will try to apply the transfer to the account balances for each chain. If the transfer doesn’t result in a negative account, it gives the go-ahead.
+
+## Detailed Design
+
+### Accountant Smart Contract
+
+#### Query methods for signature verification
+
+Since the global-accountant is a cosmwasm-based smart contract rather than a built-in native cosmos-sdk module, we will need to expose some query methods from the native wormhole module to cosmwasm (like getting the members of a guardian set).  Luckily, extending the contract interface is [fairly straightforward](https://docs.cosmwasm.com/docs/1.0/integration#extending-the-contract-interface).  The `x/wormhole` native module will expose the following query methods:
+
+```rust
+#[cw_serde]
+#[derive(QueryResponses)]
+pub enum WormholeQuery {
+    /// Verifies that `data` has been signed by a quorum of guardians from `guardian_set_index`.
+    #[returns(Empty)]
+    VerifyQuorum {
+        data: Binary,
+        guardian_set_index: u32,
+        signatures: Vec<Signature>,
+    },
+
+    /// Verifies that `data` has been signed by a guardian from `guardian_set_index`.
+    #[returns(Empty)]
+    VerifySignature {
+        data: Binary,
+        guardian_set_index: u32,
+        signature: Signature,
+    },
+
+    /// Returns the number of signatures necessary for quorum for the given guardian set index.
+    #[returns(u32)]
+    CalculateQuorum { guardian_set_index: u32 },
+}
+```
+
+#### Submitting observations to wormchain
+
+When a guardian observes a token transfer message from the registered tokenbridge on a particular chain, it will submit that observation directly to the accountant.  The contract will keep track of pending token transfers and emit an event once it has received a quorum of observations.  If a quorum of guardians submit their observations within a single tendermint block then the token transfer only needs to wait for one round of tendermint consensus and one round of guardian consensus.
+
+When a guardian submits an observation to the contract, the contract will perform several checks:
+
+- Check that the signature is valid and from a guardian in the current guardian set.
+- If a transfer with the same `(emitter_chain, emitter_address, sequence)` tuple has already been committed, then check that the digest of the observation matches the digest of the committed transfer.  If they match return a `Committed` status; otherwise return an `Error` status and emit an `Error` event.
```

Comment:
not entirely true as it could be pending / redundant

Reply:
I believe a transfer in the committed state already would have 13/19 signatures and passed the balance check, so basically only committed or error should be returned at this point.

Perhaps it would be more clear to reorder this bullet to come before this one.  

```
- If the transfer has not been committed, mark the observation as having a signature from the guardian, and return a `Pending` status
```

Also I believe transfer and observation are used interchangeably but kind of mean the same thing?  Let's replace `observation` with `transfer`.

Reply:
I believe I've implemented your suggestions @conorpp in 8abbb46a902fec028dd9f41a834a4a30765a9588. Please let me know if you disagree.